internal/leasecontroller/lease_controller.go

```diff
@@ -0,0 +1,335 @@
+package leasecontroller
+
+import (
+	"context"
+	pkgerrors "errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/virtual-kubelet/virtual-kubelet/internal/nodepingcontroller"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	"github.com/virtual-kubelet/virtual-kubelet/trace"
+	coord "k8s.io/api/coordination/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/typed/coordination/v1beta1"
+)
+
+type Status int
+
+const (
+	// Pending indicates we have not yet found entered a state that is known
+	Pending Status = iota
+	// Established means at some point, we established a lease.
+	Established
+	// FailedToEstablish indicates that we tried to establish a lease, but we failed to
+	FailedToEstablish
+)
+
+type LeaseController interface {
+	// leaseControllerStatus returns whether the Lease Controller has successfully been able to establish a lease.
+	LeaseControllerStatus(ctx context.Context) Status
+	// Run starts the lease controller. It blocks until context is cancelled, or if the lease cannot be established.
+	Run(ctx context.Context)
+}
+
+// NewV1BetaV1LeaseController creates a new lease controller.
+//
+// shouldUpdateOrEstablishLeaseFunc is a function that may block, but will be called every lease update interval,
```

Comment:
`shouldUpdateOrEstablishLeaseFunc` doesn't exist (anymore?) and it's actually `Run`.

Also, my understanding of the code below is that the `Run` function is indeed blocking on `lc.poll(ctx)` which means this function may actually not run within the lease update interval, i.e. `lc.poll(ctx)` didn't complete within lease interval.